### DISCUSSION TITLE
Meeting Minutes from December 18, 2023

### DISCUSSION THREAD

**Discussion by dezidizon** [CVWRSM/cvwrsm#35] [Meeting Minutes]
**Open Action Items:**
- Coordinate with Hao Xie for manual steps in creating WRIMS installer and patch artifacts. This is tracked as [https://github.com/CVWRSM/wrims2/issues/29](url)
- Apply copyright files (license and source) in the wrims2 repository using Eclipse Public License (EPL) version 2 or 3. This is tracked as [https://github.com/CVWRSM/wrims2/issues/16](url)
- RMA has sent a DSS global variable flag for Hao Xie to use in order to force WRIMS to write DSS6 files. RMA will help in diagnosing the issues as this is not working correctly. This is tracked as [https://github.com/CVWRSM/wrims2/issues/31](url)
- Remove vista libraries from WRIMS. This is tracked as [https://github.com/CVWRSM/wrims2/issues/33](url)
- Document descriptions for each folder in WRIMS. [https://github.com/CVWRSM/wrims2/issues/34](url)

**Discussions:**
- Tom Evans presented the work done in preparing a README file for WRIMS to be merged into the "Feature/HecDSS7" branch of the CVWRSM/wrims2 repository. This README describes how to develop WRIMS.
    - There is a top level README file that describes the WRIMS application and links to its page on the DWR website.
    - The top level README links to the building process for WRIMS, which was the procedure Tom learned to build WRIMS.
- Tom Evans is conducting a dependency analysis to convert WRIMS and to create a master gradle build that can become the GitHub gradle actions.
    - This process begins with identifying the least dependent models. The work begins from the least dependent models to the next dependent models.
    - Tom used Idep from the Eclipse IDE which goes through jar files to examine dependencies.
    - The dependency analysis work started with:
        - gov.ca.dwr.jdiagram
        - gov.ca.dwr.hecdssvue
        - wrimsv2_plugin
        - wrimsv2
    - Tom noted that the wrims set-up is self-contained. The wrimsv2 jar does not have any external references to ther things that are built in the process and are well self-contained. There are outside references to CalSim models but are launched to system level executables rather than through Java mechanisms.
    - Hao noted that calsim, dsm2, and wresl.xtext are not built.
    - Hao requested having the vista library removed in the WRIMS DevOps effort moving to use HecDSS7. Additional libraries that need to be removed will be identified.
        - The vista source exists in [https://github.com/CADWRDeltaModeling/dsm2-vista](url)
- Bring WRIMS up to the latest HEC code but still use DSS version 6 files. The goal is to add heclib7 but continue to produce DSS version 6 files.
    - There is a global static setting during initialization to kick heclib to default to DSS version 6 and make sure it doesn't get switched.
- In another branch, Tom Evans took the contents of DSSVue. The libraries were built using the latest Eclipse IDE but the compiler targeted Java 8 b202. There were a handful of method calls that needed to be changed because deprecated functions have been removed.
    - Tom was unable to build the complete executable.
- Hao Xie wants to move on to using long-term stable releases of Java.
    - If new version of eclipse is used, a new version of Java is also needed.
    - Hao needs to decide if Eclipse should be updated to a more recent but not the latest version.
- Deziree Dizon discussed the creation of the GitHub project of "WRIMS DevOps" in the repository of CVWRSM/wrims2.
    - Tickets created for the WRIMS DevOps effort can be viewed as a list or in a Kanban board. Swim lanes are created based on milestones.
- Peter Morris and Adam Korynta discussed splitting WRIMS into separate repositories - one for the WRIMS installation and one for the plugins.
    - Peter and Adam asked to confirm that the work for WRIMS moving forward goes under the current repository of wrims2. 
        - This was previously "wrims" then renamed to "cvm". This is now "wrims2".
        - Hao will have to decide with Nazrul Islam regarding the final name of the repository.
    - Because there is ongoing work in the wrims2 repository, the source code for the plugins is intended to be forked into a different repository but still under the same group. 
    - All plug-ins that need to be built must be ensured. And parts that are no longer needed moving forward must be removed.
    - Adjustments to plugins can be done independently from WRIMS parts.
- Peter Morris recommended documenting details of the existing folders in the current repository of CVWRSM/wrims2. 
    - RMA will work with Hao to develop the descriptions for each folder.
    - Each folder will have their own README file with the description. These will be collected in a README in the main repository.
- Can Dogrul discussed the effort for migrating the Integrated Water Flow Model (IWFM) from Subversion to GitHub.
    - IWFM is written in Fortran.
    - Can intends for RMA to work on the migration process while staying heavily involved with the process.
    - IWFM may be brought into GitHub as its own organization. Multiple repositories may be created depending on how IWFM is set-up and versioning.
        - IWFM currently has 10 repositories.
        - Peter recommended considering what DWR's plans are for GitHub organizations as there are enterprise-only features. This will be posed to Nazrul Islam.
    - Peter will prepare a written plan for migrating IWFM from Subversion to GitHub. The plan will include the following items:
        - The history for IWFM in Subversion will be brought into GitHub.
        - Dependency analysis will be done for IWFM before conversion to Git.
        - A mapping device will be used and parts of the project will be routed to identify whether to organize IWF under one or multiple repositories.
        - The mapping will be tested and reviewed with Can.
    - Can will trust that Git will have the master copy of IWFM once he is comfortable working with GitHub.
    - Can expects that the repository can be sent to RMA and may have access to the entire history of the project.
        - The repository may be archived and sent to RMA.
    - There is active development in IWFM.
    - Nicky Sandhu noted that IWFM has no jars. This may be handled differently for Gradle compared to the effort for WRIMS.
    - Can noted that the migration work for IWFM is not urgent and can occur after effort for WRIMS is completed.
- Adam Korynta discussed how to handle multiple versions of artifacts in the same repository. This can be difficult to manage.
    - The version can be tracked in the parent build tool file. This is onerous in the Git workflow because the version number will have to be updated.
    - In all HEC repositories now, the version is based on the branch name. The version is picked up instead of being managed in the build tool file. But for multiple artifacts, all those artifacts will have the version number.
    - The version number itself goes into the artifact management system and not the artifact binary produced.
    - For IWFM, MS Build or Cmake will be used to figure out what's needed for Fortran. A version number will be needed for the artifact.
- Adam Korynta described the refactor from the current feature branch into the new repository in the establishment of Gradle to build one module, including that Gradle build into artifact management. This is discussing the separation of the wrims_v2 and wrimsv2_plugin.
    - Peter recommended that Hao takes the opportunity in this migration effort to rename modules as he sees fit and use standard naming.
    -  The wrims_v2 would be migrated into a "wrims_v2_core" directory with its entire history. Adam wants to know what other build tools would have to be built in.
        - Hao recommends keeping wrims_v2 instead of being renamed to "wrims_v2_core". The wrimsv2_plugin may be renamed.
    - All the modules for wrims_v2 would be placed at the root directory.
    - For wrimsv2_plugin, the OSGI plugin will be built on Gradle and tested with the existing system. This would result in a baseline test. Then the rest of the plugin can be  pulled in.
    - Multiple plugins can be done simultaneously. Once all the plugins are built, then the first release of WRIMS can be produced.
    - The home for the artifacts must be identified. Should these be in a publicly accessible DWR repository or should these go on mavencentral or other github packages.

**Closed Action Items:**
- Coordinate a follow-up meeting to cover work already done for the WRIMS DevOps effort.
- RMA will determine then propose strategy for merging in work for WRIMS DevOps updates into current repository under wrims2.

**Comment by rma-psmorris** [comment]
Additions to notes: 

- For WRESL Xtext -- Hao will look at how this dependency works
- There is an Ant build file for WRIMS v2.
- Hao eventually wants WRIMS to move to newer version of Java 
- @adamkorynta -- Can you please respond with your ideas and coupling between module naming and group id, artifact name? I don't see enough coverage of that in Deziree's notes.

**Comment by hxiedwr** [comment]
xtext.editor.ui depends on xtext.editor and wrimsv2_plugin.

Hao
________________________________

**Comment by adamkorynta** [comment]
Notes on module naming:

Currently the Eclipse projects use the reversed fully qualified domain name (similar to java packages) for the folder names. This provides nice distinctions to the modules so that devs can decipher what is an Eclipse project vs other directories. 

For maven artifact management, a set of coordinates is used: groupId, artifactId, and version. Gradle's convention is to use the artifactId as the directory name for the module.

When moving the plugins into a separate repository I suggest the following:

- Set the groupId to `gov.ca.water`
- Set the artifactIds to dash delimited, lower case versions of the current projects. For examples:
  - calsim-hydro
  - hec-dssvue
  - jdiagram
  - wresl-feature-site
  - wresl-feature
  - wresl-xtext-editor-tests
  - wresl-xtext-editor-ui
  - wresl-xtext-editor
  - wrims-calsimshp
  - wrims-geoschematic
  - wrims-v2
  - wrims-v2-plugin

Additionally, the ./wrims_v2/wrims_v2 and ./wrims_v2/wrimsv2_plugin directories should be moved up to the top level of the repository. 

Defining the purpose of each module and mapping out module dependencies will help to identify if any of the existing Eclipse projects should become consolidated or should be renamed further.

**Comment by hxiedwr** [comment]
I would suggest a name of wrims-ide instead of wrims-v2-plugin as the name of wrims v2 plugin is too general.
________________________________

**Comment by rma-tom** [comment]
The dependency analysis tool I used is named "jdeps"
